[Attributor] Don't replace `addrspacecast (ptr null to ptr addrspace(x))` with `ptr addrspace(x) null`

shiltian opened the pull request:
`ConstantPointerNull` represents a pointer with value 0, but it doesn’t
necessarily mean a `nullptr`. `ptr addrspace(x) null` is not the same as
`addrspacecast (ptr null to ptr addrspace(x))` if the `nullptr` in AS X is not
zero. Therefore, we can't simply replace it.

Fixes #115083.

arsenm commented on llvm/lib/Transforms/IPO/AttributorAttributes.cpp:
```diff
@@ -10970,6 +10970,15 @@ struct AAPotentialValuesImpl : AAPotentialValues {
+      // doesn’t necessarily mean a nullptr. `ptr addrspace(x) null` is not the
+      // same as `addrspacecast (ptr null to ptr addrspace(x))` if the nullptr
+      // in AS x is not zero. Therefore, we can't simply replace it.
+      if (isa<ConstantPointerNull>(NewV) && isa<ConstantExpr>(OldV)) {
+        auto &CE = cast<ConstantExpr>(OldV);
```
dyn_cast instead of isa + null

shiltian replied:
yeah, but here we want to check both and only use the name for the second one.

arsenm replied:
You can use the C++17 `if (auto *CE = dyn_cast<>; CE && CE->getOpcode() != ...)`

arsenm commented on llvm/lib/Transforms/IPO/AttributorAttributes.cpp:
```diff
@@ -10970,6 +10970,15 @@ struct AAPotentialValuesImpl : AAPotentialValues {
+      // doesn’t necessarily mean a nullptr. `ptr addrspace(x) null` is not the
+      // same as `addrspacecast (ptr null to ptr addrspace(x))` if the nullptr
+      // in AS x is not zero. Therefore, we can't simply replace it.
+      if (isa<ConstantPointerNull>(NewV) && isa<ConstantExpr>(OldV)) {
```
Does this handle nested subexpressions correctly?

shiltian replied:
The result type must be a pointer so that leaves us to `IntToPtr` and `AddrSpaceCast` for the outermost expression. If it is `IntToPtr`, I think they can be directly replaced because the value will have to be 0. With that being said, I think this would work fine.

arsenm replied:
Specifically there are some stripPointerCasts that should be restricted to stripPointerCastsSameRepresentation

shiltian replied:
Interestingly, that was exactly the first version of my fix, but I chose the current approach after a second thought.

We still want to look through address space cast when analyzing potential pointer value and/or underlying object. The only thing concerned here is probably nullptr. I think it is okay to treat `ptr addrspace(x) null` same as `addrspacecast (ptr null to ptr addrspace(x))` when we do analysis such that we can know whether a pointer points to nothing, but we will not replace the latter with the former because of the limitation/issue/design we currently have.

On a side note, IMHO, they should be the same and `ptr addrspace(x) null` should present the nullptr in AS x but the source code representation `ConstantPointerNull` means something else. Once we have a new class for DL nullptr, I think `ptr addrspace(x) null` should belong to it.

nikic replied:
> We still want to look through address space cast when analyzing potential pointer value and/or underlying object.

It's okay to do this for underlying objects, because the semantics allow it. But I don't think it's valid for PotentialValues.

shiltian commented on llvm/test/Transforms/Attributor/do-not-replace-addrspacecast-with-constantpointernull.ll:
```diff
@@ -0,0 +1,39 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt -S -passes=attributor $s -o - | FileCheck %s
```
for some reason it work... :-D

shiltian commented:
* **#126779** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/126779?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/126779?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

shiltian commented:
Ping.

AFAICT, only `ConstantPointerNull` matters here because it represents `null` in all AS. For the others, they are not constant thus will not be replaced by this code path.

arsenm commented on llvm/lib/Transforms/IPO/AttributorAttributes.cpp:
```diff
@@ -10970,6 +10970,15 @@ struct AAPotentialValuesImpl : AAPotentialValues {
+      // doesn’t necessarily mean a nullptr. `ptr addrspace(x) null` is not the
+      // same as `addrspacecast (ptr null to ptr addrspace(x))` if the nullptr
+      // in AS x is not zero. Therefore, we can't simply replace it.
+      if (isa<ConstantPointerNull>(NewV)) {
+        if (auto *CE = dyn_cast<ConstantExpr>(&OldV);
+            CE && CE->getOpcode() == Instruction::AddrSpaceCast)
+          continue;
+      }
```
Checking this here, in manifest, and for this specific expression doesn't feel like the right way to address this.

I would trust it slightly more if it was a check on the address space of the pointer type, but I would still expect this logic to be somewhere else

arsenm replied:
Can this sink somewhere down in getSingleValue?

arsenm commented on llvm/test/Transforms/Attributor/do-not-replace-addrspacecast-with-constantpointernull.ll:
```diff
@@ -0,0 +1,18 @@
+  %select = select i1 %icmp, ptr %p0, ptr null
+  %load = load i32, ptr %select, align 4
+  ret i32 %load
+}
```
Also add another test with a differently typed null

arsenm replied:
Also test vector case ?

shiltian replied:
added one, but we can't differentiate them (at the moment) so even if it has 0 null, it will not be replaced.

shiltian commented:
This will be properly resolved after https://github.com/llvm/llvm-project/pull/131557 (as well as its follow-up patches).

shiltian commented:
I figured it's gonna take much longer time to change the semantics of `ConstantPointerNull` as well as all related stuff, so I pushed an alternative solution here.

arsenm commented on llvm/test/Transforms/Attributor/do-not-replace-addrspacecast-with-constantpointernull.ll:
```diff
@@ -0,0 +1,34 @@
+  %select = select i1 %icmp, ptr %p0, ptr null
+  %load = load i32, ptr %select, align 4
+  ret i32 %load
+}
```
Also add the versions with addrspace(1) instead of 5 which will use 0 init

arsenm replied:
But you should, that would make the test more useful

shiltian replied:
Added, though honestly I don't think it matters at this point. We don't currently have a way to notify the middle end whether a `nullptr` in a given address space is zero or not, and there's no middle-end hook for that either, so it won't actually change anything in practice. I'd be surprised if it made any difference right now.

arsenm replied:
But it will catch the change in the future patch. There's no plus to not making this a target test that will expose the difference

shiltian commented on llvm/test/Transforms/Attributor/do-not-replace-addrspacecast-with-constantpointernull.ll:
```diff
@@ -0,0 +1,34 @@
+; CHECK-LABEL: define i32 @foo(
+; CHECK-SAME: ptr nofree readonly captures(none) [[P0:%.*]], ptr nofree noundef nonnull readonly align 16 captures(none) dereferenceable(16) [[P1:%.*]], <2 x ptr addrspace(5)> [[PTRVEC:%.*]]) #[[ATTR0]] {
+; CHECK-NEXT:    [[LOADVEC:%.*]] = load <2 x ptr addrspace(5)>, ptr [[P1]], align 16
+; CHECK-NEXT:    [[ICMPVEC:%.*]] = icmp eq <2 x ptr addrspace(5)> [[LOADVEC]], <ptr addrspace(5) addrspacecast (ptr null to ptr addrspace(5)), ptr addrspace(5) addrspacecast (ptr null to ptr addrspace(5))>
```
`addrspacecast (<2 x ptr> zeroinitializer to <2 x ptr addrspace(5)>)` gets converted to `<ptr addrspace(5) addrspacecast (ptr null to ptr addrspace(5)), ptr addrspace(5) addrspacecast (ptr null to ptr addrspace(5))>` automatically even w/o running anything.

jdoerfert reviewed (approved):
I'm not super convinced but it seems necessary for now.
LG

shiltian committed:
[Attributor] Don't replace `AddrSpaceCast` with `ConstantPointerNull`

`ConstantPointerNull` represents a pointer with value 0, but it doesn’t
necessarily mean a nullptr. `ptr addrspace(x) null` is not the same as
`addrspacecast (ptr null to ptr addrspace(x))` if the nullptr in AS x is not
zero. Therefore, we can't simply replace it.

Fixes #115083.
